lib/active_record/connection_adapters/postgresql_fallback_adapter.rb

```diff
@@ -0,0 +1,102 @@
+require 'active_record/connection_adapters/abstract_adapter'
+require 'active_record/connection_adapters/postgresql_adapter'
+require 'discourse'
+require 'concurrent'
+
+class PostgreSQLFallbackHandler
+  include Singleton
+
+  class TaskObserver
+    def update(time, result, ex)
+      if result
+        logger.info { "PG connection heartbeat successfully returned #{result}" }
+      elsif ex.is_a?(Concurrent::TimeoutError)
+        logger.warning { "PG connection heartbeat timed out".freeze }
+      else
+        if ex.message.include?("could not connect to server")
+          logger.info { "PG connection heartbeat: Master connection is not active.".freeze }
+        else
+          logger.error { "PG connection heartbeat failed with error: \"#{ex}\"" }
+       end
+      end
+    end
+
+    private
+
+    def logger
+      Rails.logger
+    end
+  end
+
+  def start(existing_connection, config)
+    return if @running
+    @running = true
+    Discourse.enable_readonly_mode
+
+    timer_task = Concurrent::TimerTask.new(execution_interval: interval) do |task|
```

Comment:
I am a bit uneasy with adding another thread here, it feels too heavy, instead hijack when you checkout a connection like the connection reaper does